In [1]:
# Importing the required libraries
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline


In [2]:
# Reading the csv file and putting it into 'df' object.
df = pd.read_csv("M:\Imarticus\python project\crop-varietal-identification-with-scio\Barley.data.csv")
df.head()

Predictor       740       741       742       743       744       745  \
0     HB-52  1.499853  1.500665  1.501564  1.502536  1.503537  1.504548   
1     HB-52  1.530877  1.531669  1.532524  1.533419  1.534308  1.535169   
2     HB-52  1.607175  1.608642  1.610044  1.611357  1.612540  1.613585   
3     HB-52  1.677333  1.678098  1.678930  1.679809  1.680688  1.681539   
4     HB-52  1.507830  1.509293  1.510683  1.511977  1.513145  1.514180   

        746       747       748    ...         1061      1062      1063  \
0  1.505588  1.506689  1.507828    ...     1.636411  1.637350  1.636452   
1  1.536024  1.536916  1.537830    ...     1.643299  1.644222  1.643302   
2  1.614534  1.615456  1.616359    ...     1.703343  1.704291  1.703329   
3  1.682384  1.683260  1.684145    ...     1.755067  1.756013  1.754996   
4  1.515125  1.516049  1.516964    ...     1.641968  1.642877  1.641946   

       1064      1065      1066      1067      1068      1069      1070  
0  1.634370  1.632287  1.630265  1.630722  1.633166  1.636204  1.637872  
1  1.641195  1.639088  1.637046  1.637493  1.639938  1.642981  1.644649  
2  1.701137  1.698946  1.696822  1.697278  1.699806  1.702953  1.704677  
3  1.752716  1.750441  1.748238  1.748696  1.751290  1.754525  1.756294  
4  1.639829  1.637714  1.635664  1.636101  1.638535  1.641567  1.643227  

[5 rows x 332 columns]

In [3]:
#collecting indepedent variables in X
X = df.iloc[:,1:332]
X_col = X.columns
X.head()

740       741       742       743       744       745       746  \
0  1.499853  1.500665  1.501564  1.502536  1.503537  1.504548  1.505588   
1  1.530877  1.531669  1.532524  1.533419  1.534308  1.535169  1.536024   
2  1.607175  1.608642  1.610044  1.611357  1.612540  1.613585  1.614534   
3  1.677333  1.678098  1.678930  1.679809  1.680688  1.681539  1.682384   
4  1.507830  1.509293  1.510683  1.511977  1.513145  1.514180  1.515125   

        747       748       749    ...         1061      1062      1063  \
0  1.506689  1.507828  1.508937    ...     1.636411  1.637350  1.636452   
1  1.536916  1.537830  1.538712    ...     1.643299  1.644222  1.643302   
2  1.615456  1.616359  1.617205    ...     1.703343  1.704291  1.703329   
3  1.683260  1.684145  1.684969    ...     1.755067  1.756013  1.754996   
4  1.516049  1.516964  1.517836    ...     1.641968  1.642877  1.641946   

       1064      1065      1066      1067      1068      1069      1070  
0  1.634370  1.632287  1.630265  1.630722  1.633166  1.636204  1.637872  
1  1.641195  1.639088  1.637046  1.637493  1.639938  1.642981  1.644649  
2  1.701137  1.698946  1.696822  1.697278  1.699806  1.702953  1.704677  
3  1.752716  1.750441  1.748238  1.748696  1.751290  1.754525  1.756294  
4  1.639829  1.637714  1.635664  1.636101  1.638535  1.641567  1.643227  

[5 rows x 331 columns]

In [4]:
#collecting depedent variable in Y
Y=df['Predictor']

Pre-processing methods are helpful in eliminating noise generated by spectral data. Raw spectral data were thus processed using a combination of scatter corrections that include Standard Normal Variate (SNV) as well as first and second degree derivatives. The Savitzky-Golay and Gap-segment derivative smoothing filtering algorithms also usefull in eliminating noise.

Here I am using Savitzky-Golay filter with second degree derivative.

In [29]:
#Savitzky-Golay filter with second degree derivative.
from scipy.signal import savgol_filter 

sg=savgol_filter(X,window_length=11, polyorder=3, deriv=2, delta=1.0)

In [6]:
sg_x=pd.DataFrame(sg, columns=X_col)

sg_x.head()

740       741       742       743       744       745       746  \
0  0.000084  0.000073  0.000061  0.000050  0.000039  0.000028  0.000010   
1  0.000033  0.000027  0.000021  0.000014  0.000008  0.000001 -0.000011   
2 -0.000139 -0.000128 -0.000117 -0.000105 -0.000094 -0.000082 -0.000078   
3  0.000055  0.000043  0.000031  0.000020  0.000008 -0.000004 -0.000022   
4 -0.000148 -0.000133 -0.000119 -0.000105 -0.000091 -0.000077 -0.000069   

        747       748       749    ...         1061      1062      1063  \
0 -0.000011 -0.000030 -0.000044    ...    -0.000446 -0.000474 -0.000300   
1 -0.000024 -0.000036 -0.000042    ...    -0.000446 -0.000474 -0.000300   
2 -0.000073 -0.000069 -0.000060    ...    -0.000462 -0.000491 -0.000310   
3 -0.000043 -0.000061 -0.000071    ...    -0.000471 -0.000502 -0.000315   
4 -0.000062 -0.000055 -0.000045    ...    -0.000446 -0.000473 -0.000299   

       1064      1065      1066      1067      1068      1069      1070  
0  0.000008  0.000319  0.000781  0.001243  0.001705  0.002167  0.002629  
1  0.000010  0.000322  0.000786  0.001250  0.001714  0.002178  0.002642  
2  0.000010  0.000334  0.000815  0.001296  0.001777  0.002258  0.002738  
3  0.000014  0.000347  0.000842  0.001337  0.001832  0.002327  0.002822  
4  0.000010  0.000322  0.000786  0.001249  0.001713  0.002176  0.002640  

[5 rows x 331 columns]

In [8]:
# Splitting the data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =train_test_split(sg, Y,
                                                    test_size=0.2,
                                                    random_state=123,stratify=Y)

In [9]:
from sklearn.ensemble import RandomForestClassifier

Rf = RandomForestClassifier(random_state=52)
Rf_fit=Rf.fit(X_train, y_train)

In [10]:
y_pred = Rf.predict(X_test)

In [11]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print("Test Result:\n")        
print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Classification Report: \n {}\n".format(classification_report(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test,y_pred))) 

Test Result:

accuracy score: 0.8083

Classification Report: 
               precision    recall  f1-score   support

 Ardu 1260 B       0.89      0.80      0.84        10
      Bahati       0.53      0.90      0.67        10
    Bekoji-1       1.00      1.00      1.00        10
 Cross 41/98       0.91      1.00      0.95        10
      Deribe       0.70      0.70      0.70        10
       Dimtu       1.00      1.00      1.00        10
     EH 1493       1.00      0.80      0.89        10
     EH 1847       0.80      0.80      0.80        10
    Explorer       0.75      0.90      0.82        10
       Grace       0.67      0.60      0.63        10
     HB-1307       0.75      0.90      0.82        10
     HB-1533       0.62      0.50      0.56        10
     HB-1963       1.00      0.90      0.95        10
     HB-1964       1.00      1.00      1.00        10
     HB-1965       0.80      0.80      0.80        10
     HB-1966       0.62      0.50      0.56        10
 IBON 174/03      

In [30]:
#Reduction of variables using Recursive Feature Elimination(RFE) techineque

from sklearn.feature_selection import RFE

# RFE with 10 features

rfe_10 = RFE(Rf,10)

rfe_10.fit(X_train, y_train)

# Printing the boolean results
print(rfe_10.support_)             

[False False False False False False False False False False False False
 False False False False  True False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False  True False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False  True False False False False False
 False False False False False False  True False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [14]:
y_pred = rfe_10.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred))) 

accuracy score: 0.7250

Confusion Matrix: 
 [[ 8  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  8  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0]
 [ 0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  1  0  9  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  2  0  0  7  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  9  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  9  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0]
 [ 0  0  0  0  2  0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  2  0  0  0  0  1  0  0  7  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  1  0  0  0  8  0  1  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  1  0  0  0  0  0  0  5  0  0  0  0  0  1  0  0  0  0  3  0]
 [ 0  0  0  0  2  0  1  0  0  0  0  0  7  0  0  0  0  0 

In [15]:
# RFE with 15 features
rfe_15 = RFE(Rf,15)

# fit with 15 features
rfe_15.fit(X_train, y_train)

# Printing the boolean results
print(rfe_15.support_)           


[False False False False False False False False False False False False
 False False False False  True False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False  True  True False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False  True False False False False False
 False False False False False False  True False False False  True False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [17]:
y_pred = rfe_15.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred))) 

accuracy score: 0.7292

Confusion Matrix: 
 [[ 8  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  9  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  1  0  7  0  0  0  1  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  2  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  2  0  0  0  0  6  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  6  0  0  0  0  0  0  0  0  0  0  4  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  2  0  0  0  0  2  0  0  6  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  9  0  0  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  1  0  0  0  0  0  0  0  5  0  0  0  0  0  0  0  0  0  0  4  0]
 [ 0  0  0  0  2  0  0  0  0  0  0  0  8  0  0  0  0  0 

In [22]:
# RFE with 17 features

rfe_17 = RFE(Rf,17)

# fit with 17 features
rfe_17.fit(X_train, y_train)

# Printing the boolean results
print(rfe_17.support_)      

[False False False False False False False False False False False False
 False False False False  True False False  True False False False False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False  True  True False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False  True  True False False False False
 False False False False False False  True False False False  True False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [23]:
y_pred = rfe_17.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred))) 

accuracy score: 0.7750

Confusion Matrix: 
 [[ 8  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  7  0  0  0  0  1  0  0  0  0  0  0  1  0  0  0  0  0  0  1  0  0  0]
 [ 0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0]
 [ 0  0  1  0  7  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  2  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  7  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  1  0  0  0  0  2  0  0  7  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  9  0  0  0  0  0  0  0  0  0  0  1  0]
 [ 0  0  0  0  2  0  0  0  0  0  0  0  8  0  0  0  0  0 

In [24]:
# RFE with 20 features

rfe_20 = RFE(Rf,20)

# fit with 20 features
rfe_20.fit(X_train, y_train)

# Printing the boolean results
print(rfe_20.support_)

[False False False False False False False False False False False False
 False False False False  True False False  True False False False False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False  True  True False False False False False False False False  True
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False  True  True False False False False
 False False False False False False  True False False False  True False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [26]:
y_pred = rfe_20.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred))) 

accuracy score: 0.7250

Confusion Matrix: 
 [[ 8  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  6  0  0  0  0  1  0  0  1  0  0  0  0  0  0  1  0  0  0  1  0  0  0]
 [ 0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  9  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  0  0  0  5  1  0  0  2  0  0  0  2  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  8  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0]
 [ 0  0  0  0  4  0  0  0  5  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  1  0  0  0  0  1  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  9  0  0  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  1  0  0  0  0  0  0  0  5  0  0  0  0  0  0  0  0  0  0  4  0]
 [ 0  0  0  0  1  0  0  0  0  0  0  0  9  0  0  0  0  0 

In [28]:
Rf.fit(X_train, y_train)
# Get numerical feature importances
importances = list(Rf.feature_importances_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 4)) for feature, importance in zip(X_col, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances 
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances]

Variable: 757                  Importance: 0.0161
Variable: 790                  Importance: 0.0141
Variable: 1059                 Importance: 0.014
Variable: 848                  Importance: 0.0106
Variable: 1065                 Importance: 0.0105
Variable: 975                  Importance: 0.0094
Variable: 1043                 Importance: 0.0094
Variable: 750                  Importance: 0.0091
Variable: 752                  Importance: 0.0091
Variable: 1064                 Importance: 0.0091
Variable: 960                  Importance: 0.009
Variable: 1039                 Importance: 0.0084
Variable: 783                  Importance: 0.0081
Variable: 991                  Importance: 0.0081
Variable: 1046                 Importance: 0.0081
Variable: 854                  Importance: 0.0079
Variable: 843                  Importance: 0.0078
Variable: 1015                 Importance: 0.0075
Variable: 1017                 Importance: 0.0073
Variable: 844                  Importance: 0.0072
Va

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

conclusion:- Random Forest is giving accuracy of 80% using all 331 variables. If we reduce Variables using techineque Recursive Feature Elimination(RFE),using 17 variables model is giving acurracy of 77%. Further increase in Variables will not affecting model accuracy as much.